File: python/vec3.py
from random import random
from math import sqrt


class Vec3:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def __iter__(self):
        return iter((self.x, self.y, self.z))

    def __truediv__(self, num):
        return Vec3(self.x / num, self.y / num, self.z / num)

    def __mul__(self, num):
        return Vec3(self.x * num, self.y * num, self.z * num)

    def __rmul__(self, num):
        return Vec3(self.x * num, self.y * num, self.z * num)

    def __add__(self, v):
        return Vec3(self.x + v.x, self.y + v.y, self.z + v.z)

    def __sub__(self, v):
        return Vec3(self.x - v.x, self.y - v.y, self.z - v.z)


def unit(v):
    return v / length(v)


def length_squared(v):
    return v.x * v.x + v.y * v.y + v.z * v.z


def length(v):
    return sqrt(length_squared(v))


def dot(a, b):
    return a.x * b.x + a.y * b.y + a.z * b.z


def random_in_unit_sphere():
    result = Vec3(
        (random() - 0.5) * 2.0,
        (random() - 0.5) * 2.0,
        (random() - 0.5) * 2.0,
    )
    if length_squared(result) >= 1:
        result.x = (random() - 0.5) * 2.0
        result.y = (random() - 0.5) * 2.0
        result.z = (random() - 0.5) * 2.0
    return result


def reflect(v, normal):
    return unit(v - normal * 2.0 * dot(v, normal))

File: zig/src/vec3.zig
const std = @import("std");
const math = std.math;
const Random = std.rand.Random;

pub const Vec3 = @Vector(3, f32);

pub inline fn toVec3(value: f32) Vec3 {
    return @splat(3, value);
}

pub inline fn dot(a: Vec3, b: Vec3) f32 {
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
}

pub inline fn lenSquared(v: Vec3) f32 {
    return v[0] * v[0] + v[1] * v[1] + v[2] * v[2];
}

pub inline fn len(v: Vec3) f32 {
    return math.sqrt(lenSquared(v));
}

pub inline fn unit(v: Vec3) Vec3 {
    return v / toVec3(len(v));
}

pub inline fn randomInUnitSphere(rnd: *const Random) Vec3 {
    var result = Vec3{
        (rnd.float(f32) - 0.5) * 2.0,
        (rnd.float(f32) - 0.5) * 2.0,
        (rnd.float(f32) - 0.5) * 2.0,
    };
    while (lenSquared(result) >= 1) {
        result[0] = (rnd.float(f32) - 0.5) * 2.0;
        result[1] = (rnd.float(f32) - 0.5) * 2.0;
        result[2] = (rnd.float(f32) - 0.5) * 2.0;
    }
    return result;
}

pub inline fn reflect(a: Vec3, n: Vec3) Vec3 {
    return a - n * toVec3(2.0) * toVec3(dot(a, n));
}
